gdx/src/com/badlogic/gdx/graphics/g2d/GlyphLayout.java

```diff
@@ -94,201 +112,164 @@ public void setText (BitmapFont font, CharSequence str, Color color, float targe
-				if (runEnd != runStart) { // Can occur eg when a color tag is at text start or a line is "\n".
-					// Store the run that has ended.
-					GlyphRun run = glyphRunPool.obtain();
-					run.color.set(color);
-					fontData.getGlyphs(run, str, runStart, runEnd, lastGlyph);
-					if (run.glyphs.size == 0) {
-						glyphRunPool.free(run);
-						break runEnded;
-					}
-					int xAdvancesCount = run.xAdvances.size;
-					float[] xAdvances = run.xAdvances.items;
-					GlyphRun previous = null;
-					if (lastGlyph != null) { // Move back the width of the last glyph from the previous run.
-						previous = runs.peek();
-						x += xAdvances[0] - previous.xAdvances.peek();
-						xAdvances[0] = 0;
-					}
-					run.x = x;
-					run.y = y;
-					runs.add(run);
-
-					if (newline || runEnd == end) {
-						adjustLastGlyph(fontData, run);
-						lastGlyph = null;
-					} else
-						lastGlyph = run.glyphs.peek();
-
-					if (!wrapOrTruncate || xAdvancesCount == 0) { // No wrap or truncate, or no glyphs.
-						if (markupEnabled) { // If disabled, any subsequent run is sure to be on the next line.
-							for (int i = 0; i < xAdvancesCount; i++)
-								x += xAdvances[i];
-						}
-						break runEnded;
+			runEnded:
+			{
+				// Store the run that has ended.
+				GlyphRun run = glyphRunPool.obtain();
+				run.x = 0;
+				run.y = y;
+				fontData.getGlyphs(run, str, runStart, runEnd, lastGlyph);
+				glyphCount += run.glyphs.size;
+
+				if (nextColor != currentColor) { // Can only be different if markupEnabled.
+					if (colors.get(colors.size - 2) == glyphCount) {
+						// Consecutive color changes, or after an empty run, or at the beginning of the string.
+						colors.set(colors.size - 1, nextColor);
+					} else {
+						colors.add(glyphCount);
+						colors.add(nextColor);
 					}
+					currentColor = nextColor;
+				}
+
+				if (run.glyphs.size == 0) {
+					glyphRunPool.free(run);
+					if (lineRun == null) break runEnded; // Otherwise wrap and truncate must still be processed for lineRun.
+				} else if (lineRun == null) {
+					lineRun = run;
+					runs.add(lineRun);
+				} else {
+					lineRun.appendRun(run);
+					glyphRunPool.free(run);
+				}
 
-					// Wrap or truncate.
-					x += xAdvances[0]; // First xadvance is the first glyph's X offset relative to the drawing position.
-					for (int i = 1; i < xAdvancesCount; i++) {
-						Glyph glyph = run.glyphs.get(i - 1);
+				if (newline || isLastRun) {
+					setLastGlyphXAdvance(fontData, lineRun);
+					lastGlyph = null;
+				} else
+					lastGlyph = lineRun.glyphs.peek();
+
+				if (!wrapOrTruncate || lineRun.glyphs.size == 0) break runEnded; // No wrap or truncate, or no glyphs.
+
+				if (newline || isLastRun) {
+					// Wrap or truncate. First xadvance is the first glyph's X offset relative to the drawing position.
+					float runWidth = lineRun.xAdvances.first() + lineRun.xAdvances.get(1); // At least the first glyph will fit.
+					for (int i = 2; i < lineRun.xAdvances.size; i++) {
```

Comment:
~~@Nathan Sweet: As there was a change to `i = 2` here, shouldn't it also be `i = 2;` in line 242?
Testing look great with `i = 2;` in line 242 but I guess just because I don't have a font with a really wide single charater (> 3 * spaceWidth)... or do I miss something?~~

@NathanSweet Nevermind the above comment, `i = 1;` yields the correct result while wrapping in the BitmapFontTester (`i = 2;` does not, it wraps "too late" for some lines), although I don't 100% understand why.

Reply:
It's because right after the `i = 1;` the loop repeats, so `i` is incremented and is 2 for the next go around.